geo/src/algorithm/triangulate_delaunay.rs

```diff
@@ -0,0 +1,645 @@
+        let poly = Polygon::new(self.into_iter().cloned().collect(), vec![]);
+        poly.delaunay_triangulation()
+    }
+}
+
+fn create_super_triangle<T: GeoFloat>(geometry: &Polygon<T>) -> Result<Triangle<T>>
+where
+    f64: From<T>,
+{
+    let expand_factor = T::from(DEFAULT_SUPER_TRIANGLE_EXPANSION)
+        .ok_or(DelaunayTriangulationError::FailedToConstructSuperTriangle)?;
+    let bounds = geometry
+        .bounding_rect()
+        .ok_or(DelaunayTriangulationError::FailedToConstructSuperTriangle)?;
+    let width = bounds.width() * expand_factor;
+    let height = bounds.height() * expand_factor;
+    let bounds_min = bounds.min();
+    let bounds_max = bounds.max();
+
+    Ok(Triangle(
+        coord! {x: bounds_min.x - width, y: bounds_min.y},
+        coord! {x: bounds_max.x + width, y: bounds_min.y - height},
+        coord! {x: bounds_max.x + width, y: bounds_max.y + height},
+    ))
+}
+
+fn find_line<T: GeoFloat>(line: &Line<T>, lines: &[Line<T>]) -> Option<usize> {
+    for (idx, sample_line) in lines.iter().enumerate() {
+        if line == sample_line {
+            return Some(idx);
+        }
+    }
+    None
+}
+
+fn add_coordinate<T: GeoFloat>(triangles: &mut Vec<DelaunayTriangle<T>>, c: &Coord<T>)
+where
+    f64: From<T>,
+{
+    let mut bad_triangles: Vec<&DelaunayTriangle<T>> = Vec::new();
+
+    // Check for the triangles where the point is present within the
+    // corresponding circumcircle.
+    for tri in triangles.iter() {
+        if tri.is_in_circumcircle(c) {
+            bad_triangles.push(tri);
+        }
+    }
+
+    let mut bad_triangle_edges: Vec<Line<T>> = Vec::new();
+    let mut bad_triangle_edge_count: Vec<u128> = Vec::new();
+    for x in &bad_triangles {
+        for y in x.0.to_lines() {
+            let flipped_y = Line::new(y.end, y.start);
+            let idx =
+                find_line(&y, &bad_triangle_edges).or(find_line(&flipped_y, &bad_triangle_edges));
+            if let Some(idx) = idx {
+                // The unwrap is acceptable due to the push lines below.
+                let count = bad_triangle_edge_count.get_mut(idx).unwrap();
+                *count += 1;
+            } else {
+                // These two vectors must be updated at the same time
+                bad_triangle_edges.push(y);
+                bad_triangle_edge_count.push(1);
+            }
+        }
+    }
+
+    // Shared edges are those with a count of > 1
+    let polygon: Vec<Line<T>> = bad_triangle_edge_count
+        .iter()
+        .enumerate()
+        .filter(|(_, count)| **count < 2)
+        .map(|(idx, _)| bad_triangle_edges[idx])
+        .collect();
+
+    // Remove the bad triangles
+    let mut new_triangles: Vec<DelaunayTriangle<T>> = triangles
+        .iter()
+        .filter(|x| !bad_triangles.contains(x))
+        .cloned()
+        .collect();
+
+    polygon
+        .iter()
+        .for_each(|x| new_triangles.push(DelaunayTriangle(Triangle(x.start, x.end, *c))));
```

Comment:
Not sure, but: Can't we inline `polygon` here and not collect above into a new `Vec`?

Edit: Ah just saw that we push to the `new_triangles`. Then please try this instead to avoid needless allocation (you'll probably have to add a `.cloned()` to the iterator somewhere.

```rust
   // Shared edges are those with a count of > 1
    let polygon = bad_triangle_edge_count
        .iter()
        .enumerate()
        .filter(|(_, count)| **count < 2)
        .map(|(idx, _)| bad_triangle_edges[idx]); // <- no collect here anymore

    // Remove the bad triangles
    let mut new_triangles: Vec<DelaunayTriangle<T>> = triangles
        .iter()
        .filter(|x| !bad_triangles.contains(x))
        .cloned()
        .chain(polygon) // <- maybe polygon.cloned()
        .collect();
```